File: suffix_Array/C++/find_suffix_Array.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <utility>

using namespace std;

string text;

void find_suffix_array() {
    int n = text.length();
    vector<pair<int, string>> myList;

    for (int a = 0; a < n; a++) {
        myList.push_back({a, text.substr(a)});
    }

    // Sort the list based on suffixes
    sort(myList.begin(), myList.end(), 
         [](const auto& a, const auto& b) { return a.second < b.second; });

    // Print the sorted suffix array map
    cout << "Suffix Array:" << endl;
    for (const auto& answer : myList) {
        cout << answer.first << " ";
    }
    cout << endl;
}

int main() {
    cin >> text;
    find_suffix_array();
    return 0;
} 

File: suffix_Array/Kotlin/find_suffix_Array.kt
var text: String = ""

fun find_suffix_array() {
    val n = text.length
    val myList = mutableListOf<Pair<Int, String>>()

    for (a in 0 until n) {
        myList.add(Pair(a, text.substring(a)))
    }

    // Sort the list based on suffixes
    myList.sortBy { it.second }

    // Print the sorted suffix array map
    println("Suffix Array:")
    for (answer in myList) {
        print("${answer.first} ")
    }
    println()
}

fun main() {
    text = readLine() ?: ""
    find_suffix_array()
} 